### DISCUSSION TITLE
Commissioning Devices in Matter.js: Order of Operations?

### DISCUSSION THREAD

**Discussion by joehecn** [matter-js/matter.js#1022] [Q&A] [answered]
Hey everyone,

I'm finding the examples in ControllerNode.ts and DeviceNode.ts somewhat perplexing. 
My main question is: 
Do I need to initially add the device to the network by employing the commissioningController.commissionNode method before I'm able to establish a device node using the ServerNode.create method?

Thanks for your assistance!

**Comment by Apollon77** [comment, marked as answer]
Please do not mix Device and Controller because these are different topics!

So for each device you create one ServerNode instance using the ServerNode.create method, then you add your needed endpoints to it (Root endpoint is already included in ServerNode) and then you run() it or use bringOnline().
The "Network" is included in the Environment initialized in the background and the node is automatically added there. you can also startup multiple ServerNodes by using different ports.

The Controller is encapsulated by the CommissionigController class which gets the used (or default) environment added as startup parameters. This is because the Device API is already renewed, the controller one not yet - that's why the work a bit differently. And a controller is just a client and connects to device nodes.

So in fact Device and Controller are two completely different roles in the matter ecosystem, so what you need is about what you want to do:
* A device offers a functionality like the available devices out there like a plugin socket or a lamp
* A controller is the administrator of a Matter fabric like Apple Home or Google Home

So normally you do one of the two and not both ;-)

But both examples should show how to use the relevant parts. And yes you can also have a controller and devices in the same NodeJs process.

If you have questions please ask and we will try to support the best we can.

I also convert this question into a discussion.

**Comment by joehecn** [comment]
Thank you for your reply; I now have a clearer understanding of the distinctions between ServerNode and ClientNode.
